src/solvers/dgmulti/shock_capturing.jl

```diff
@@ -0,0 +1,284 @@
+                                        volume_flux_fv, dg, cache, alpha_element)
+  end
+
+  return nothing
+end
+
+get_sparse_operator_entries(i, j, mesh::DGMultiMesh{1}, cache) =
+  SVector(cache.sparse_hybridized_SBP_operators[1][i, j])
+
+function get_sparse_operator_entries(i, j, mesh::DGMultiMesh{2}, cache)
+  Qr, Qs = cache.sparse_hybridized_SBP_operators
+  return SVector(Qr[i, j], Qs[i, j])
+end
+
+function get_sparse_operator_entries(i, j, mesh::DGMultiMesh{3}, cache)
+  Qr, Qs, Qt = cache.sparse_hybridized_SBP_operators
+  return SVector(Qr[i, j], Qs[i, j], Qt[i, j])
+end
+
+get_contravariant_matrix(element, mesh::DGMultiMesh{1}, cache) =
+  SMatrix{1, 1}(cache.dxidxhatj[1, 1][1, element])
+
+function get_contravariant_matrix(element, mesh::DGMultiMesh{2, <:Affine}, cache)
+  (; dxidxhatj) = cache
+  return SMatrix{2, 2}(dxidxhatj[1, 1][1, element], dxidxhatj[2, 1][1, element],
+                       dxidxhatj[1, 2][1, element], dxidxhatj[2, 2][1, element])
+end
+
+function get_contravariant_matrix(element, mesh::DGMultiMesh{3, <:Affine}, cache)
+  (; dxidxhatj) = cache
+  return SMatrix{3, 3}(dxidxhatj[1, 1][1, element], dxidxhatj[2, 1][1, element], dxidxhatj[3, 1][1, element],
+                       dxidxhatj[1, 2][1, element], dxidxhatj[2, 2][1, element], dxidxhatj[3, 2][1, element],
+                       dxidxhatj[1, 3][1, element], dxidxhatj[2, 3][1, element], dxidxhatj[3, 3][1, element])
+end
+
+# computes an algebraic low order method with internal dissipation.
+# TODO: implement for curved meshes
+function low_order_flux_differencing_kernel!(du, u, element, mesh::DGMultiMesh,
+                                             have_nonconservative_terms::False, equations,
+                                             volume_flux_fv, dg::DGMultiFluxDiff{<:GaussSBP},
+                                             cache, alpha=true)
+
+  # accumulates output from flux differencing
+  rhs_local = cache.rhs_local_threaded[Threads.threadid()]
+  fill!(rhs_local, zero(eltype(rhs_local)))
+
+  u_local = view(cache.entropy_projected_u_values, :, element)
+
+  (; sparsity_pattern) = cache
+  A_base = parent(sparsity_pattern) # the adjoint of a SparseMatrixCSC is basically a SparseMatrixCSR
+  row_ids, rows = axes(sparsity_pattern, 2), rowvals(A_base)
+  for i in row_ids
+    u_i = u_local[i]
+    du_i = zero(u_i)
+    for id in nzrange(A_base, i)
+      j = rows[id]
+      u_j = u_local[j]
+
+      # compute (Q_1[i,j], Q_2[i,j], ...) where Q_i = ∑_j dxidxhatj * Q̂_j
+      geometric_matrix = get_contravariant_matrix(element, mesh, cache)
+      reference_operator_entries = get_sparse_operator_entries(i, j, mesh, cache)
+      normal_direction_ij = geometric_matrix * reference_operator_entries
+
+      n_ij_norm = norm(normal_direction_ij)
+      f_ij = volume_flux_fv(u_i, u_j, normal_direction_ij / n_ij_norm, equations)
+      du_i = du_i + 2 * f_ij * n_ij_norm
```

Comment:
Do you need to normalize the direction? I thought we implemented everything such that it doesn't assume normalized directions.

Reply:
You're right! Good point, removing that.